src/sage/schemes/hyperelliptic_curves/jacobian_generic.py

```diff
@@ -152,3 +152,79 @@ def _point_homset(self, *args, **kwds):
 
     def _point(self, *args, **kwds):
         return jacobian_morphism.JacobianMorphism_divisor_class_field(*args, **kwds)
+
+    ####################################################################
+    # Endomorphism rings and related properties
+    ####################################################################
+
+    def endomorphism_ring(self):
+        """
+        The ring of endomorphisms of this Jacobian which are defined over
+        the base field.
+
+        EXAMPLES::
+
+            sage: R.<x> = QQ[]
+            sage: f = x^6 + 78*x^5 + 16*x^2 + x + 8
+            sage: C = HyperellipticCurve(f)
+            sage: J = C.jacobian()
+            sage: E = J.endomorphism_ring()
+            sage: E
+            Endomorphism ring of Jacobian of Hyperelliptic Curve over Rational Field defined by y^2 = x^6 + 78*x^5 + 16*x^2 + x + 8
+        """
+
```

Comment:
Here I would also use `lazy_attribute`

Reply:
Again as above this seems to change the functionality, the user would now do `J.endomorphism_ring` rather than `J.endomorphism_ring()` which seems to go against standard sage usage

Reply:
I didn't explain myself well enough.
Here is what I meant:
```
@lazy_attribute
def __endomorphism_ring(self):
    from .jacobian_endomorphisms import EndomorphismRing
    return EndomorphismRing(self)
	
def endomorphism_ring(self):
    return self.__endomorphism_ring
```

Reply:
Thanks for clarifying -- done.